devito/types/dense.py

```diff
@@ -1454,3 +1453,99 @@ def parent(self):
         return self._parent
 
     _pickle_kwargs = Function._pickle_kwargs + ['parent']
+
+
+class TempFunction(DiscreteFunction):
+
+    is_TempFunction = True
+
+    def __init_finalize__(self, *args, **kwargs):
+        super().__init_finalize__(*args, **kwargs)
+
+        self._pointer_dim = kwargs.get('pointer_dim')
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        if pointer_dim not in dimensions:
+            # This is a bit hacky but it does work around duplicate dimensions when
+            # it gets to pickling
+            dimensions = as_tuple(pointer_dim) + dimensions
+        return dimensions, dimensions
+
+    def __halo_setup__(self, **kwargs):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        halo = as_tuple(kwargs.get('halo'))
+        if halo is None:
+            halo = tuple((0, 0) for _ in dimensions)
+        if pointer_dim is not None:
+            halo = ((0, 0),) + halo
+        return halo
+
+    @property
+    def data(self):
+        # Any attempt at allocating data by the user should fail miserably
+        raise TypeError("TempFunction cannot allocate data")
+
+    data_domain = data
+    data_with_halo = data
+    data_ro_domain = data
+    data_ro_with_halo = data
+
+    @property
+    def pointer_dim(self):
+        return self._pointer_dim
+
+    @property
+    def dim(self):
```

Comment:
why the need for both?

Reply:
Make it usable where we have a PointerArray, as it's `PointerArray.dim`. It'd be cleaner with a Mixin, I admit... I may add one